libraries/lib-wave-track/WaveClip.cpp

```diff
@@ -1146,6 +1160,65 @@ void WaveClip::Resample(int rate, BasicUI::ProgressDialog *progress)
    }
 }
 
+void WaveClip::ApplyStretchRatio()
+{
+   const auto stretchRatio = GetStretchRatio();
+   if (stretchRatio == 1.0)
+      return;
+
+   Finally Do { [this] {
+      this->mClipStretchRatio = 1.0;
+      this->mRawAudioTempo = this->mProjectTempo;
+   } };
+
+   constexpr auto durationToDiscard = 0.0;
+   constexpr auto blockSize = 1024;
+   const auto numChannels = GetWidth();
+
+   // Let's also pass the hidden part to the stretcher for better result in the
+   // visible part.
+   const auto trimLeftBeforeStretch = mTrimLeft;
+   const auto trimRightBeforeStretch = mTrimRight;
+   SetTrimLeft(0);
+   SetTrimRight(0);
+
+   ClipSegment stretcherSource { *this, durationToDiscard,
+                                 PlaybackDirection::forward };
+   TimeAndPitchInterface::Parameters params;
+   params.timeRatio = stretchRatio;
+   StaffPadTimeAndPitch stretcher { numChannels, stretcherSource,
+                                    std::move(params) };
+   const auto totalNumOutSamples =
+      sampleCount { GetVisibleSampleCount().as_double() * stretchRatio + .5 };
+   sampleCount numProcessedSamples { 0 };
+
+   auto newSequences = GetEmptySequenceCopies();
+
+   sampleCount numOutSamples { 0 };
+   AudioContainer container(blockSize, numChannels);
+   while (numOutSamples < totalNumOutSamples)
+   {
+      const auto numSamplesToGet = limitSampleBufferSize(
+         blockSize, totalNumOutSamples - numProcessedSamples);
+      stretcher.GetSamples(container.Get(), numSamplesToGet);
+      auto channel = 0u;
+      for (auto& newSequence : newSequences)
+         newSequence->Append(
+            reinterpret_cast<samplePtr>(container.Get()[channel++]),
+            floatSample, numSamplesToGet, 1,
+            widestSampleFormat /* computed samples need dither */
+         );
+      numOutSamples += numSamplesToGet;
+   }
+
+   SetTrimLeft(trimLeftBeforeStretch * stretchRatio);
+   SetTrimRight(trimRightBeforeStretch * stretchRatio);
+
+   mSequences = move(newSequences);
+   Flush();
```

Comment:
`Flush` might fail and throw.  So instead, swap `newSequences` into place.  In the `Finally`, if there is not success, then swap again.

Reply:
To restore the original sequences I'd have the `Flush` in a try-catch, so that I can detect if it's failed. But it seems to me that you're suggestion something different? Also, there's another such a `Flush` call that's not handled (I copy-pasted from it) ; is the situation there different?